Update graphs for OpenShift

spadgett opened the pull request:
* Add HOC that determines whether to show components that require promtheus
* Restore charts to cluster overview page for OpenShift
* Hide Grafana link on OpenShift

/assign @alecmerdler

spadgett commented:
@alecmerdler Opinion on this approach? Trying to find a clean way to decide when to show charts

alecmerdler commented on frontend/public/components/graphs/index.jsx:
```diff
@@ -15,3 +16,40 @@ export const Gauge = props => <AsyncComponent loader={() => import('./graph-load
+};
+
+// HOC that will hide WrappedComponent when Prometheus is not configured or the user doesn't have permission.
+export const withGraphs = WrappedComponent => connectToFlags(FLAGS.OPENSHIFT, FLAGS.PROMETHEUS, FLAGS.CAN_LIST_NS)(props => {
```
Seems like a good approach to me.

spadgett commented:
jenkins push

coreos-ui commented:
The PR image has been pushed to Quay with tag `pr_183_build_396`

jhadvig commented on frontend/public/components/cluster-overview.jsx:
```diff
@@ -25,126 +25,128 @@ const fetchHealth = () => coFetch(`${k8sBasePath}/healthz`)
-        <h2 className="h3">Control Plane Status</h2>
-        <DashboardLink to="/grafana/dashboard/db/kubernetes-control-plane-status?orgId=1" />
+        <h2 className="h3">Health</h2>
+        {!isOpenShift && <DashboardLink id="qa_dashboard_k8s_health" to="/grafana/dashboard/db/kubernetes-cluster-health?orgId=1" />}
```
Since the `to` path is used on multiple places, with different fragments, wouldn't it be better to construct it with a function ?

spadgett replied:
I'm a little hesitant to change this since I don't have a good way to test.

I kind of prefer that they're opaque URLs, though. It's not clear they'll always follow the same pattern.

jhadvig replied:
yeah I understand your concern. On the other hand if they change the pattern it would be more easy if the path is stored in constant.
But that could be done as a followUp :)

jhadvig commented on frontend/public/components/cluster-overview.jsx:
```diff
@@ -25,126 +25,128 @@ const fetchHealth = () => coFetch(`${k8sBasePath}/healthz`)
+
+const LimitedGraphs = requirePrometheus(({isOpenShift}) => {
+  const consoleTitle = getConsoleTitle(isOpenShift);
+  return <React.Fragment>
```
No need for the `<React.Fragment>` since there is only one child.

jhadvig commented on frontend/public/components/cluster-overview.jsx:
```diff
@@ -8,12 +8,12 @@ import { NavTitle, AsyncComponent, Firehose, StatusBox, tectonicHelpBase, OpenSh
 import { SoftwareDetails } from './software-details';
 import { formatNamespacedRouteForResource } from '../ui/ui-actions';
-import { FLAGS, connectToFlags } from '../features';
+import { connectToFlags, FLAGS } from '../features';
```
Not sure if this change is needed

spadgett replied:
Yeah, that was unintentional (although I do prefer to alphabetize imports)

jhadvig commented on frontend/public/components/graphs/index.jsx:
```diff
@@ -15,3 +16,40 @@ export const Gauge = props => <AsyncComponent loader={() => import('./graph-load
+  return canListNS;
+};
+
+// HOC that will hide WrappedComponent when Prometheus isn't configured or the user doesn't have permission to query Prometheus.
```
:+1:

jhadvig commented on frontend/public/components/namespace.jsx:
```diff
@@ -141,6 +141,29 @@ class PullSecret extends SafetyFirst {
+      ]} />
+    </div>
+  </div>
+  <Bar title="Memory Usage by Pod (Top 10)" query={`sort(topk(10, sum by (pod_name)(container_memory_usage_bytes{pod_name!="", namespace="${ns.metadata.name}"})))`} humanize={humanizeMem} metric="pod_name" />
```
Those `query`'s are a bit harder to read. Wondering if there is a way how to make it easier.

spadgett replied:
Yeah, just moving existing code here.

It might be possible to format them across lines.

jhadvig replied:
yeah, was thinking out loud... no need to do it now

jhadvig commented on frontend/public/components/nav.jsx:
```diff
@@ -338,6 +338,7 @@ export class Nav extends React.Component {
           <NavSection text="Troubleshooting" icon="fa-life-ring">
             <HrefLink href="/search" name="Search" onClick={this.close} startsWith={searchStartsWith} />
             <ResourceNSLink resource="events" name="Events" onClick={this.close} />
+            { /* FIXME: These links are wrong for OpenShift. */ }
```
Should we open a story for that ?

spadgett replied:
Yeah, I've been using CONSOLE-585 to track getting metrics to work in OpenShift generally. Will add to that story. There are a few other related problems.

spadgett replied:
I ended up opening a separate issue for this so we don't lose track (CONSOLE-602)

spadgett commented:
Thanks @jhadvig 

I removed the unnecessary React.Fragment and fixed the import.

I'd prefer not to change the existing dashboard URLs or queries.

jhadvig commented:
@spadgett yeah makes sense since there is not an easy way how to test. Thanks :+1:
